react/features/base/devices/services/device-detect/ActiveDeviceDetector.js

```diff
@@ -0,0 +1,143 @@
+// @flow
+import EventEmitter from 'events';
+import { ACTIVE_DEVICE_DETECTED } from './Events';
+import logger from '../../logger';
+import JitsiMeetJS from '../../../lib-jitsi-meet';
+
+const JitsiTrackEvents = JitsiMeetJS.events.track;
+
+// If after 3000 ms the detector did not find any active devices consider that there aren't any usable ones available
+// i.e. audioLevel > 0.008
+const DETECTION_TIMEOUT = 3000;
+
+/**
+ * Detect active input devices based on their audio levels, currently this is very simplistic. It works by simply
+ * checking all monitored devices for TRACK_AUDIO_LEVEL_CHANGED if a device has a audio level > 0.008 ( 0.008 is
+ * no input from the perspective of a JitsiLocalTrack ), at which point it triggers a ACTIVE_DEVICE_DETECTED event.
+ * If there are no devices that meet that criteria for DETECTION_TIMEOUT an event with empty deviceLabel parameter
+ * will be triggered,
+ * signaling that no active device was detected.
+ * TODO Potentially improve the active device detection using rnnoise VAD scoring.
+ */
+export class ActiveDeviceDetector extends EventEmitter {
+
+    /**
+     * Currently monitored devices.
+     */
+    _availableDevices: Array<Object>;
+
+    /**
+     * State flag, check if the instance was destroyed.
+     */
+    _destroyed: boolean = false;
+
+    /**
+     * Create active device detector.
+     *
+     * @param {Array<MediaDeviceInfo>} micDeviceList - Device list that is monitored inside the service.
+     *
+     * @returns {ActiveDeviceDetector}
+     */
+    static async create(micDeviceList: Array<MediaDeviceInfo>) {
```

Comment:
Having this included in the constructor has a similar problem to https://github.com/jitsi/lib-jitsi-meet/pull/982#discussion_r348374937 . 
I understand that we could mask the loading of async objects in the c'tor or some other initialise like function, but we would still have to deal with the async nature at one point or another, because we have a cleanup method here that expects the async objects to be loaded when it's called. Of course you could change the `destroy` method to wait for the objects to load but that would mean you need to handle the async nature of `destroy`. IMO having to deal with a async factory method is cleaner than having to deal with that state inside the class itself. But again that's just me if you have a more compelling reason, I'll can change it.